### DISCUSSION TITLE
iRobot Create 3 IMU Coordinate Frames

### DISCUSSION THREAD

**Discussion by Paras-05** [iRobotEducation/create3_docs#389] [Questions]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Computer(s) Model(s) and Operating System(s)

Raspberrypi 4 / Ubuntu 20.04

### Which version of ROS 2 is installed on your computer?

Galactic

### Which firmware version is installed on your robot?

G.4.5

### Which RMW is your robot running?

FastRTPS

### Is the robot connected to a network? If so, what is the network type?

Modem (Laptop, Raspi, Robot on the same network)

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

USB

### Describe your question.

Hello everyone,

**Background:** I am working with iRobot Create 3 to implement 3d visual inertial odometry using TOF camera input and the IMU data input.

**System & Network:** I have connected Raspi 4 with the Create3 through USB. To establish the simple network connectivity, I am using modem which connects the Raspi, Create3, Laptop. Through ssh into the Raspi, I am building my algorithms in it.

**ROS Middleware:** As my algorithms are written in ROS1. I am using ros-bridge package to bridge all the create3 ROS2 topics. Basically, I am building my whole environment in ROS1 using all input data from ros_bridge (/imu, /tf, /camera). 

**Problems:** 
1. How should I know the position of all co-ordinate frames on the create3 robot - /odom, /base_link, /base_footprint, /imu physically so that I can write the transformations for the /camera_link and /imu_link wrt /base_link or /base_footprint or /odom in my launch file. what are the packages should I incorporate in my project wrt this problem.

2. Is the IMU used in the create3 robot is calibrated. or is there any need to add filters.

Please share any links, documentation, and your expertise in this regard.

Sincerely,
Paras

**Comment by justinIRBT** [comment]
1) The base_link/base_footprint coordinate frame is in the middle of the robot, between the two wheels, on the ground.  base_link factors in the IMU pitch/roll, base_footprint is projected onto the ground.  I would use the distance between your camera and the center of the robot on the ground (z = 0).  The other robot sensors (cliff, bump, wheels, etc...) are with respect to that location as well.  You can just use a StaticTransformBroadcaster to publish this relationship for your custom sensors to get it into the TF tree.
2) The Create3 IMU is calibrated and does have bias removal built in.  There are tricks you can use to keep the heading accuracy for longer periods of time, for example if you rotate a lot in one direction, you can remove some of the error accumulated by rotating in the other direction, for example, if you rotate 90 degrees 3 times (270 total), to return to 0, instead of rotating another 90, you can rotate -270.  There is a RotateAngle action that closes the loop on gyro heading if you want to understand gyro error from accumulated rotation: https://iroboteducation.github.io/create3_docs/api/drive-goals/

**Comment by Paras-05** [comment]
Hello Justin,

Hope you have a good day!

Thanks a lot for the reply.

I would be great if you answer some of my queries regarding the same discussion:

1. According to the side view pic of Create3 (attached), what height should I consider if I am attaching my camera from the top faceplate of the robot.
![side_view](https://github.com/iRobotEducation/create3_docs/assets/82272724/e71d7ec5-da9e-43d7-b0e7-e3eead292751)

2. According to the top view (faceplate removed), what is the center point of the robot so that I would consider some x distance for my camera.
![Top view with faceplate removed](https://github.com/iRobotEducation/create3_docs/assets/82272724/d17764f0-6fd9-43d1-aaaa-0f424d3a06d9)

3. Is the base link and the base footprint are on the same coordinate frame. Is there any difference in between the both to be considered.
It would be quite helpful if you write me the exact coordinate position (x, y, z) of the base link, the base footprint, and /imu with respect to base_link.

Thanks for your time and consideration.

Regards,
Paras

**Comment by justinIRBT** [comment]
1) You are using height from floor to top of faceplate, so 93.4mm in the side view drawing, then add the height of your camera's focal point from the faceplate.

2) center point is the intersection between the primary x-axis and y-axis dotted lines (the lines that all measurements are relative to that visibly appear to intersect the vertical and horizontal center of the circle)

3) I am not sure I understand the question here.  The base_link / base_footprint represent the location on the robot with respect to other coordinate frames (frequently `odom` and `map`).  They are both at the same point (center of the robot on the ground).  So when the robot starts up and drives 1m forward for example, your `odom->base_link` transform will be (1.0,0.0,0.0) and the position it is talking about on the robot is the center on the ground.  You can then query your `odom->camera_link` which is attached to `base_link`, so it will have the (1.0,0.0,0.0) plus the transform offset you have from base_link to get the position of that sensor in the `odom` coordinate system.  The difference between `base_link` and `base_footprint` is that `base_footprint` is projected onto the ground plane, whereas `base_link` accounts for the pitch/roll in the IMU.  This difference would be most prominent if you were driving up a ramp, where the linear travel would differ between the 2 given that one had non-trivial Z translation, whereas base_footprint would project all motion onto the XY plane, so the x/y coordinates would show more movement.  In general, if you are doing something with a camera, I would recommend to attach to base_link to account for the pitch/roll of the robot.  `base_footprint` is frequently used with lasers that are doing planar SLAM.

The IMU is offset from robot center:
IMU x offset  = 0.06
IMU y offset  = 0.035
IMU z offset = 0.0685

**Reply by Paras-05** [reply]
Hello Justin,

Thanks for your informative reply.

I have set up all the required transformations for the Create3 wrt my application. In my launch file, I have added the following lines wrt each coordinate origin:

![hj](https://github.com/iRobotEducation/create3_docs/assets/82272724/3f4a116f-15b2-412a-bcf1-5e496aa4d793)

Please guide me if I am going somewhere wrong with these transformations.
Attached you will find the flow chart generated with /tf.

![jk](https://github.com/iRobotEducation/create3_docs/assets/82272724/f087dec9-856f-4ad8-ae47-42b6f657f75a)

Thanks for your time.

Regards,
Paras

**Reply by justinIRBT** [reply]
This connection hierarchy looks correct.  One correction from my previous response, the imu offset on the x axis is 0.051 instead of 0.06.  A future Create3 release will publish the imu frame in its tf tree.  A couple of notes:

- One thing to double check is that the optical frame doesn't have a rotation.  Sometimes you see optical frames that use a y down, x to the right coordinate system.  If your device produces a point cloud, you can confirm this by looking at it in rviz relative to the base_link frame.  An example with this rotated optical frame is realsense (copied from a launch file):
```
Node(package='tf2_ros', executable='static_transform_publisher',
            arguments=['0.0', '0.03', '0.0', str(-math.pi/2), '0', str(-math.pi/2), 'camera_link', 'camera_depth_optical_frame']),
```
- Currently you are making a static transform for your `map->odom` frame.  In the future your slam system will be in charge of producing the transform that maintains this `map->odom` relationship.

**Reply by Paras-05** [reply]
Hello Justin,

Thanks for the quick reply and quality information.

1. In regards to the info provided,  
    I have listened to the tf for the pico_flexx_optical_frame wrt pico_flexx_link. I have noticed that there is a rotation to the down. I have 
    attached the rviz representation of both the links and the transformation of pico_flexx_optical_frame wrt to the pico_flexx_link.
    
    
 
    
![rviz pico_flexx_link](https://github.com/iRobotEducation/create3_docs/assets/82272724/6640af04-7176-4830-a7e8-4fc07db84237)


![rviz pico_flexx_optical_frame](https://github.com/iRobotEducation/create3_docs/assets/82272724/1a3618c5-0e3e-4c87-88ec-9c2510f7c340)

  
![pico_optical_frame wrt pico_link](https://github.com/iRobotEducation/create3_docs/assets/82272724/25d92bf0-9f01-411e-9fea-dbf6e0c3ec17)

2. If I listened to the tf for the pico_flexx_optical_frame wrt base_link, I got the following attached results:


![pico_optical_frame wrt base_link](https://github.com/iRobotEducation/create3_docs/assets/82272724/6e34deea-6790-4217-b485-b3d3bfdb04c9)

Note: the transformation of the pico_flexx_optical_frame wrt the pico_flexx_link is predefined by the camera manufacturer.

Question: Is this change has to be written in the launch file OR writing transformation of the pico_flexx_link wrt the base_link works. If there is need to write this transformation in the launch file, could you suggest me how should I write it in my launch file.

Thanks for your time and consideration.

Regards,
Paras Sikka

**Reply by justinIRBT** [reply]
What are you running right now that is publishing the pico_flexx_link and pico_flexx_optical_frame?  The frames look fine, although, given that there is no translation offset between the link and the optical frame, that means you will want the link to be at the focal point of the camera.  Not sure if your measurements are to that location.  You can confirm it is happy by viewing the point cloud relative to base_link and confirming the location of features in the point cloud is where you expect (ground is 0, a clear feature, like a box in the point cloud has the correct x/y measurement relative to the center of the robot, etc..

Here is an example launch with a sensor and frame:
```
from launch import LaunchContext
from launch import LaunchDescription
from launch_ros.actions import Node
from ament_index_python.packages import get_package_share_directory
import math

def generate_launch_description():
    return LaunchDescription([
        Node(package='tf2_ros', executable='static_transform_publisher',
            arguments=['-0.03', '0.0', '0.19', '0', '0.03', '0.0', 'base_link', 'camera_link']),
        Node(package='tf2_ros', executable='static_transform_publisher',
            arguments=['0.0', '0.03', '0.0', str(-math.pi/2), '0', str(-math.pi/2), 'camera_link', 'camera_depth_optical_frame']),
        Node(
            package='custom_realsense',
            executable='realsense_pointcloud_node'
        ),
    ])
```

**Comment by Paras-05** [comment]
Hello Justin,

I hope this message finds you well.

I have a query regarding IMU sensor used in iRobot Create3. It would be very helpful if you share the datasheet of the IMU sensor used in the robot.

In my system, I am modelling IMU sensor as below:
the angular velocities by the gyroscope 𝛚𝑚=(𝜔𝑥,𝜔𝑦,𝜔𝑧)𝑇
the linear accelerations by the accelerometer 𝒂𝑚=(𝑎𝑥,𝑎𝑦,𝑎𝑧)𝑇

Modeling equations: 
𝛚𝑚=𝛚𝑟𝑒𝑎𝑙+𝛚𝑏+𝛚𝑛
𝒂𝑚=𝑹(𝒒)(𝒂𝑟𝑒𝑎𝑙−𝒈)+𝒂𝑏+𝒂𝑛
𝛚𝑛∼𝒩(𝟎,𝛔2𝜔) ; 𝒂𝑛∼𝒩(𝟎,𝛔2𝑎) ; 𝒂˙𝑏=𝒂𝑏𝑛∼𝒩(𝟎,𝛔2𝑎𝑏) ; 𝛚˙𝑏=𝛚𝑏𝑛∼𝒩(𝟎,𝛔2𝛚𝑏)

where 𝛚𝑟𝑒𝑎𝑙 and 𝒂𝑟𝑒𝑎𝑙 represent the true angular velocity and acceleration, respectively; 𝛚𝑛 and 𝒂𝑛 refer to the additive noises of the sensor which follow the Gaussian distributions in nature; and the bias part 𝛚𝑏 and 𝒂𝑏 can be described as random walk processes. The derivatives of the gyroscope bias 𝛚𝑏𝑛 = 𝛚˙𝑏 and the accelerometer bias 𝒂𝑏𝑛 = 𝒂˙𝑏 also follow the Gaussian distributions.
The error variance 𝛔2𝜔 and 𝛔2𝑎 usually found in the datasheet of the respective IMU sensor. 
Could you provide me with the necessary values.

Appreciate your time and consideration.

Regards,
Paras

**Reply by justinIRBT** [reply]
All Create 3 robots have been manufactured with the Bosch BMI088 to date
